Fixed deprecated params output

socratesgorilla opened the pull request:
Closes #6780

Commit:
Fixed deprecation not outputting multiple times

Commit:
Fixed deprecated params only outputting once #6780

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1255873/) on b54c1ab wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af38e9f0b6976df8d0a90b1c6b339bef1dbc9b00`

friedmud commented:
I'm not sure this solves the problem.  Sure - it lets the deprecated message come out once - but it also makes it come out every time that deprecated object is used.

For instance: in your test, if you use that object twice in the same input file you would see the deprecation message twice.

This also makes the stack trace only come out once (even for multiple deprecated parameters).

The larger fix here is difficult.  You need to keep track of which deprecated parameters have had a warning printed about them... and make sure to only print them once.

friedmud reviewed (changes requested):
Does not completely solve the problem - see above comment.

friedmud commented:
Note: you can't use `mooseDoOnce()` in this solution because it is set with a static variable... which will be the same for many parameters.  It really does need something more sohpisticated here.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1255933/) on a16c776 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22897/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1255946/) on a16c776 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>af38e9</th>
      <th colspan=4>#22897 a16c77</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.47%</td>
      <td>84.48%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80918</td>
      <td>80929</td>
      <td>+11</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14872</td>
      <td>14867</td>
      <td>-5</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22897/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22897/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22897/coverage/xfem)

</details>



This comment will be updated on new commits.

Commit:
Added check for if a deprecated param has already printed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1256464/) on f0d5c66 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af38e9f0b6976df8d0a90b1c6b339bef1dbc9b00`

friedmud committed:
Add the beginning of a test that uses multiple deprecated parameters

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1258051/) on 5402879 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1f76e400b05ee27576b21438fff474b78d3b9772`

Commit:
Added custom evaluation script for duplicate deprecated param test

Co-Authored-By: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

socratesgorilla commented on framework/include/parser/Parser.h:
```diff
@@ -243,6 +243,9 @@ class Parser : public ConsoleStreamInterface, public hit::Walker
   std::ostringstream * _current_error_stream;
 
+  /// Tracks whether a deprecated param has had its warning message printed already.
+  std::unordered_set<std::string> _deprec_param_tracker;
```
@permcody Here are the C++ changes at least -- Not 100% sure this is the best choice of container here or if there's a way to make this more elegant in general

socratesgorilla replied:
I figured unordered_set because it doesn't need to be sorted, just iterated, and I believe average case complexity is faster than vector?

permcody replied:
Your first priority is to get something that works and this does that. If it's in a time-sensitive area, then you can take a deeper look at optimizing something like container types. For what it's worth, vectors and arrays are by far the fastest containers for pretty much all but the most diabolical cases (like repeatedly inserting or removing in the middle). `unordered_set`s are surprisingly fast, almost always faster than just plain `set` even in cases where you'd expect `set`s underlying data structure to really fly.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1262251/) on 38adb5b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 82cb1fbdfc8d797688fe2ab4796e5de33fb143c6`

friedmud commented on framework/src/parser/Parser.C:
```diff
@@ -1206,6 +1206,13 @@ Parser::extractParams(const std::string & prefix, InputParameters & p)
+      // Check if we have already printed the deprec param message.
+      // If we haven't, add it to the tracker, and print it.
+      if (!_deprec_param_tracker.count(it.first))
+        if (p.attemptPrintDeprecated(it.first))
```
Do we really need to know if it was printed or not?  If it wasn't... then it won't be in the future... right?

socratesgorilla replied:
I'm confused what you mean here. All that happens here is that it checks if the message has been printed for that variable.

friedmud reviewed (dismissed):
This is looking really good - just one question.  Thanks for working this out!

GiudGiud commented on test/tests/misc/deprecation/tests:
```diff
@@ -62,4 +62,12 @@
+    input = 'deprecated_param.i'
+    custom_evaluation_script = 'deprec_param_custom_eval.py'
+    issues = '#6780'
+    requirement = 'The system shall output one warning per deprecated parameter, regardless of how many deprecated params are in the system.'
```
```suggestion
    requirement = 'The system shall output one warning per deprecated parameter, regardless of how many deprecated parameters are in the system.'
```

GiudGiud commented on test/tests/misc/deprecation/tests:
```diff
@@ -62,4 +62,12 @@
+    issues = '#6780'
+    requirement = 'The system shall output one warning per deprecated parameter, regardless of how many deprecated params are in the system.'
+  []
+  
```
```suggestion
```

GiudGiud commented on test/tests/misc/deprecation/deprecated_param.i:
```diff
@@ -0,0 +1,32 @@
+    D = 5
+    E = 2
+  []
+
```
```suggestion
```

GiudGiud commented on test/tests/misc/deprecation/deprecated_param.i:
```diff
@@ -0,0 +1,32 @@
+[]
+
+[Variables/u]
+
```
```suggestion
```

GiudGiud commented on test/src/kernels/DeprecatedParamDiffusion.C:
```diff
@@ -0,0 +1,45 @@
+
+registerMooseObject("MooseTestApp", DeprecatedParamDiffusion);
+
+// MooseDocs::start
```
what is this

GiudGiud commented on test/src/kernels/DeprecatedParamDiffusion.C:
```diff
@@ -0,0 +1,45 @@
+  InputParameters params = Diffusion::validParams();
+  params.addDeprecatedParam<Real>("D", 1, "The diffusivity coefficient.", "This is deprecated");
+  params.addDeprecatedParam<Real>(
+      "E", 1, "The ediffusivity coefficient.", "This is also deprecated");
```
```suggestion
      "E", 1, "The E diffusivity coefficient.", "This is also deprecated");
```

GiudGiud commented on test/src/kernels/DeprecatedParamDiffusion.C:
```diff
@@ -0,0 +1,45 @@
+
+DeprecatedParamDiffusion::DeprecatedParamDiffusion(const InputParameters & parameters)
+  : Diffusion(parameters),
+
```
```suggestion
```

GiudGiud commented on framework/src/utils/InputParameters.C:
```diff
@@ -96,16 +96,23 @@ InputParameters::set_attributes(const std::string & name, bool inserted_only)
     }
   }
+  return false;
 }
```
```suggestion
}

```

GiudGiud commented on framework/src/parser/Parser.C:
```diff
@@ -1206,6 +1206,13 @@ Parser::extractParams(const std::string & prefix, InputParameters & p)
     {
       p.set_attributes(it.first, false);
+
+      // Check if we have already printed the deprec param message.
```
```suggestion
      // Check if we have already printed the deprecated param message.
```

GiudGiud commented on framework/include/base/MooseError.h:
```diff
@@ -223,24 +223,26 @@ mooseDeprecatedStream(S & oss, bool expired, Args &&... args)
+  mooseStreamAll(ss, args...);
+  std::string msg =
+      mooseMsgFmt(ss.str(),
                   "*** Warning, This code is deprecated and will be removed in future versions:",
```
could we moose do once ONLY that line?
That way we have all the deprecated messages, and this shared message only once

socratesgorilla replied:
Which message do you mean, the `warning this code is deprecated` one?

GiudGiud replied:
Yes

GiudGiud reviewed (changes requested):
this PR is actually quite time sensitive now that I've just deprecated 20 or so parameters in the parsed expressions. 

Would be great to merge asap

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1321927/) on f0896cb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0021c9bedaef80662ee7242294388979badcaf44`

socratesgorilla committed:
Updated warning print to only print the Warning message once on deprec param

socratesgorilla commented:
@friedmud any chance on finishing review anytime soon on this? don't want to fix the merge conflicts until its ready to go

socratesgorilla committed:
Updated deprecated file output to work with new message system

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1399755/) on b756d6b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6d8f7c976166ddb4780cd848cacb98e994315634`

socratesgorilla committed:
Updated mooseDeprecatedStream to work with new deprecated print system

GiudGiud commented on framework/include/base/MooseError.h:
```diff
@@ -238,28 +238,29 @@ mooseDeprecatedStream(S & oss, const bool expired, const bool print_title, Args
+      Threads::spin_mutex::scoped_lock lock(moose_stream_lock);
+      oss << ss.str() << std::endl;
+    };
+  };
```
need to revert this, this doesnt work

GiudGiud reviewed (changes requested):
This correctly fixes the deprecation of multiple parameters however it breaks the deprecation of objects, see my comment.

I m interested on improving that other thing, but I think it will be difficult. Let s keep this PR to parameter deprecation

Can you also please remove 2 line breaks from this?
```
/Users/giudgl/projects/moose_v4/test/tests/misc/deprecation/deprecated_param.i:13: (Kernels/diffusion/D):
The parameter 'D' is deprecated.
This is deprecated



/Users/giudgl/projects/moose_v4/test/tests/misc/deprecation/deprecated_param.i:20: (Kernels/diffusion2/E):
The parameter 'E' is deprecated.
This is also deprecated
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1412924/) on 4e3014f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22897/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 56ed1a03fcfe7969b0a13e95582a8ec0a747ef8d`

GiudGiud commented on framework/include/base/MooseError.h:
```diff
@@ -240,27 +240,37 @@ mooseDeprecatedStream(S & oss, const bool expired, const bool print_title, Args
-    else
-      libMesh::write_traceout();
+    auto hash = std::hash<std::string>{}(ss.str().substr(pos + 1));
+    auto iter = Moose::_deprec_tracker.find(hash);
```
system by system is the way to go, not a global attribute 

This is also relying on convoluted logic
"hash the part of the message before the line break" that works fine now but could be a pain to debug when it stops working

This commit should be reverted. What kind of issue are you having with the factory tracking?

socratesgorilla replied:
There is no technical issue with implementing the object tracking in the factory system per se. 

My issue is more that we have to create two different systems in two different places that don't really interact with each other in order to keep track of their deprecation outputs. Instead of doing that, I figured it would be better to have all of them stored in one singular location that doesn't differentiate between what kind of deprecated output it is, just that `mooseDeprecatedStream` was called. Doing it this means that if we add new kinds of deprecation output, the system would still work with no modifications needed, and a universal tracker leaves a smaller footprint.

Design-wise, this does seem like a better solution. If the primary issue with the universal tracker is the line break check being too brittle, then that should be fixable.

GiudGiud replied:
it s not a lot of code to duplicate here. Placing code in Moose.C is an easy solution but it s not sustainable. We need to distribute the codes between systems or concentrate it in utils if there's good redundancy.

in general we should shun global variables. There's threading issues you need to be wary of when you use global variables, it s not worth it

socratesgorilla replied:
@loganharbour

GiudGiud replied:
Let s have a look. Where is this triggered for creating objects?

GiudGiud committed:
Add test for deprecated object
Co-authored by socratesgorilla

GiudGiud committed:
Keep track of object deprecation in the Factory to avoid double-printing deprecation messages
- we want once per object TYPE. Users should understand that two objects of the same type are both deprecated if one is

GiudGiud committed:
Avoid printing backtraces on deprecated object creation

GiudGiud committed:
Restore title on deprecated parameters print
attempt to print title once was not working
fixing it did not look that good due to excessive line breaks

GiudGiud commented:
Down for a review @friedmud

GiudGiud reviewed (approved):
Approving the part that is not mine.

GiudGiud commented:
Any volunteers among our 6 code owners?
